solr/core/src/java/org/apache/solr/handler/component/QueryComponent.java

```diff
@@ -865,7 +866,19 @@ protected void mergeIds(ResponseBuilder rb, ShardRequest sreq) {
 
       // Merge the docs via a priority queue so we don't have to sort *all* of the
       // documents... we only need to order the top (rows+start)
-      final ShardFieldSortedHitQueue queue = new ShardFieldSortedHitQueue(sortFields, ss.getOffset() + ss.getCount(), rb.req.getSearcher());
+      ShardFieldSortedHitQueue queue = new ShardFieldSortedHitQueueWithSameShardCompareSkip(sortFields, ss.getOffset() + ss.getCount(), rb.req.getSearcher());
+      
+      ShardFieldSortedHitQueue reRankQueue = null;
+      int reRankDocsSize = 0;
+      if(rb.getRankQuery() != null){
+          final RankQuery rankQuery = rb.getRankQuery();
+          if(rankQuery instanceof AbstractReRankQuery){
+              reRankDocsSize = ((AbstractReRankQuery) rankQuery).getReRankDocs();
+              reRankQueue = new ShardFieldSortedHitQueueWithSameShardCompareSkip(new SortField[]{SortField.FIELD_SCORE}, Math.min(reRankDocsSize, ss.getCount()), rb.req.getSearcher());
+
+              queue = new ShardFieldSortedHitQueue(sortFields, ss.getOffset() + ss.getCount(), rb.req.getSearcher());
+          }
+      }
```

Comment:
The existing `mergeIds` method is already quite long and complex and addition of `reRankQuery` and `reRankDocSize` for intricate use together with the existing `queue` increases complexity further. I wonder though if with some strategic signature choices it might be possible to somehow abstract the re-rank related logic into a class so that from the perspective of the `mergeIds` method nothing really changes?

E.g. imagine that line 869 here did not do `new ShardFieldSortedHitQueue...` directly but instead called a small wrapper method in QueryComponent:

```
private static ShardFieldSortedHitQueue newShardFieldSortedHitQueue(SortField[] sortFields, int count, int offset, IndexSearcher searcher, RankQuery rankQuery) {
  if (rankQuery instanceof AbstractReRankQuery) {
    return new FooBarShardFieldSortedHitQueue(sortFields, count, offset, searcher, ((AbstractReRankQuery) rankQuery).getReRankDocs());
    
  } else {
    return new ShardFieldSortedHitQueue(sortFields, count + offset, searcher);
  }
}
```

and well `FooBarShardFieldSortedHitQueue` is obviously not a good class name but imagine it looked something like this

```
class FooBarShardFieldSortedHitQueue extends ShardFieldSortedHitQueue {
  final private int reRankDocsSize;
  final private ShardFieldSortedHitQueue foobarQueue;

  @Override
  public void insertWithOverflowAndI(T element, int i) {
    if (i < reRankDocsSize) {
      T droppedElement = foobarQueue.insertWithOverflow(element);
      if (dropped != null ) {
        super.insertWithOverflow(dropped);
      }
    } else {
      super.insertWithOverflow(element);
    }
  }

  @Override
  public T pop1() {
    T element = super.pop();
    if (element == null) {
      element = foobar.pop();
    }
    return element;
  }

}
```

with supporting `ShardFieldSortedHitQueue` tweaks as follows

```
public class ShardFieldSortedHitQueue extends PriorityQueue<ShardDoc> {

+ public void insertWithOverflowAndI(T element, int i) {
+   super.insertWithOverflow(element);
+ }

+ @Override // workaround for super class' pop (understandably) being final
+ public T pop1() {
+   return super.pop();
+ }

}
```

and small matching `mergeId` adjustments to call `insertWithOverflowAndI` instead of `insertWithOverflow` and `pop1` instead of `pop`.

So same logic, only structured differently in terms of code with `FooBarShardFieldSortedHitQueue` potentially also more easily testable (though I haven't though much about that as yet to be honest).

Reply:
We did not fully understand your remarks but agree that the `mergeIds( )` is too complex.
Therefore we extracted the logic of handling multiple queues to a new class.
You can see this in 47075d2.

Reply:
Thanks! Yes, a `SortedHitQueueManager` style class was the key idea, the FooBar queue within queue sketching was just one possible implementation speculation.